File: 0684-redundant-connection/0684-redundant-connection.cpp
class Solution {
public:
    int n;
    vector<int> father;
    Solution() {
        n = 1005;
        father = vector<int>(n);

        for (int i = 0; i < n; i++) {
            father[i] = i;
        }
    }

    int find(int u) {
        if (father[u] == u) {
            return u;
        }
        father[u] = find(father[u]);

        return father[u];
    } 

    void join(int u, int v) {
        u = find(u);
        v = find(v);

        if (u == v) {
            return;
        }

        father[v] = u;
    }

    bool isSame(int u, int v) {
        if (find(u) == find(v)) {
            return true;
        }

        return false;
    } 
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
       for (vector<int> edge : edges) {
           if (isSame(edge[0], edge[1])) {
               return edge;
           } else {
               join(edge[0], edge[1]);
           }
       }

        vector<int> res;
       return res;
    }
};

File: 0684-redundant-connection/0684-redundant-connection.py
class Union:
        def __init__(self):
            self.parents = list(range(1005))
        
        def find(self, u):
            parent_u = self.parents[u]
            if parent_u == u:
                return u
            else:
                return self.find(parent_u)

        def join(self, u, v):
            parent_u = self.find(u)
            parent_v = self.find(v)

            self.parents[parent_v] = parent_u

        def isSame(self, u, v):
            return self.find(u) == self.find(v)

class Solution:
    

    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        union = Union()

        for edge in edges:
            u, v = edge

            if union.isSame(u, v):
                return edge
            else:
                union.join(u, v)

            